### PR TITLE
Add retry mechanism for github release uploads

### LINKED ISSUE #29: Retry github release upload if the first attempt failed due to Github

### PR DESCRIPTION
**Pull request by sdurrheimer**
Fixes #29

### CODE REVIEW

**Comment by fabxc** [comment]
Cool! I ran into this a couple of times. There are also existing libraries for backoff behavior. Alertmanager is using this one: https://github.com/cenk/backoff

Just something to consider.

**Comment by sdurrheimer** [comment]
I've been looking for a lib, but didn't really wanted to add a dependency.

**Comment by sdurrheimer** [comment]
@grobie Any objection ?

File: cmd/release.go [hunk 1]
```diff
@@ -39,6 +47,8 @@ var releaseCmd = &cobra.Command{
 // init prepares cobra flags
 func init() {
 	Promu.AddCommand(releaseCmd)
+
+	releaseCmd.Flags().IntVar(&allowedRetry, "retry", 2, "Number of retry to perform when upload fails")
```
Comment (juliusv) [inline cmd/release.go]: "Numer of retry" -> "Number of retries"

(and end flag descriptions with a period)

Comment (juliusv) [inline cmd/release.go]: and `allowedRetry` -> `allowedRetries`

File: cmd/release.go [hunk 2]
```diff
@@ -64,14 +74,29 @@ func uploadTarball(path string, f os.FileInfo, err error) error {
...
+		err := retry.Do(func(attempt int) (bool, error) {
+			var err error
+			err = uploadTarball(fileName, path)
+			if err != nil {
+				time.Sleep(2 * time.Second)
+			}
+			return attempt < allowedRetry+1, err
+		})
+		if err != nil {
+			fmt.Printf("Upload failed after %d attempts\n", allowedRetry+1)
```
Comment (juliusv) [inline cmd/release.go]: I guess in promu, printing errors to stdout with fmt.Printf is normal?

File: cmd/release.go [hunk 3]
```diff
@@ -64,14 +74,29 @@ func uploadTarball(path string, f os.FileInfo, err error) error {
...
-			"--tag", fmt.Sprintf("v%s", info.Version),
-			"--name", fileName,
-			"--file", path)
+		err := retry.Do(func(attempt int) (bool, error) {
+			var err error
+			err = uploadTarball(fileName, path)
+			if err != nil {
+				time.Sleep(2 * time.Second)
+			}
+			return attempt < allowedRetry+1, err
```
Comment (juliusv) [inline cmd/release.go]: Your retry package already does checking of max retries (but against MaxRetries = 10, not the flag value), so the slightly different second check here is a bit odd. Why not just have one limit?

**Comment by juliusv** [comment]
TBH, I would have probably just inlined the (bit simpler) retry logic, but I guess using a separate package (or external lib) is also fine if the retry patterns become more used and more complex. I'd just go for KISS at the beginning.

File: cmd/release.go (hunk 2 above)
Comment (sdurrheimer) [inline cmd/release.go]: I guess I can improve that :)

File: cmd/release.go (hunk 3 above)
Comment (sdurrheimer) [inline cmd/release.go]: The `MaxRetries` is a hard limit to prevent over exagerating the retry mechanism which would be counter productive IMO.

**Comment by sdurrheimer** [comment]
Most of the time, things are inlined given that this tool is still small, but IMO, this one has a good chance to be re-used.

File: cmd/release.go (hunk 1 above)
Comment (sdurrheimer) [inline cmd/release.go]: Fixed

File: cmd/release.go (hunk 3 above)
Comment (juliusv) [inline cmd/release.go]: That doesn't make much sense to me TBH. If a user supplies a too large retry mechanism, either just allow that or shout a warning/error in their direction immediately. But I don't think there should be two places that check the number of retries with different logic.

**Comment by grobie** [comment]
Cool! Thanks for implementing this behavior!

I share @juliusv thought in general, keeping things simple until abstractions become necessary usually helps to keep complexity low. "has a good chance" is not a strong signal to justify abstractions. Without knowing how other use cases will look like, it's easy to build the wrong interfaces. No objections to keep the current package though.

It confused me a bit to see that there are two places where the retry attempt counter is hold, once in the package method and then once again in the caller. There also seems to be a hard limit of 10 retries which is not obvious from the flag interface.

**Comment by sdurrheimer** [comment]
TBF I haven't write this retry mechanism myself, I've been searching for a light implementation without adding a external dependency.
If it needs modifications, I will do it.

**Comment by grobie** [comment]
I guess it's more important to have this feature than to argue over style in a tool source. LGTM

**Comment by sdurrheimer** [comment]
I'm gonna remove the `MaxRetries` check in the package, and rename the `utils` folder to `util` to be consistent with `prometheus`.

**Comment by sdurrheimer** [comment]
`MaxRetries` check removed.

**Comment by grobie** [comment]
LGTM

**Pull request merged**